File: api/examples/client-example.cs
/*
 * sfvector - SQL Server Vector Search with FAISS
 * Copyright (c) 2026 sfvector contributors
 * 
 */


using System.Net.Http.Json;
using System.Text.Json;

namespace Sfvector.Api.Examples;

/// <summary>
/// C# client example for sfvector API
/// </summary>
public class SfvectorClient
{
    private readonly HttpClient _httpClient;
    private string? _token;

    public SfvectorClient(string baseUrl)
    {
        _httpClient = new HttpClient { BaseAddress = new Uri(baseUrl) };
    }

    /// <summary>
    /// Login and get JWT token
    /// </summary>
    public async Task<bool> LoginAsync(string username, string password)
    {
        var request = new { username, password };
        var response = await _httpClient.PostAsJsonAsync("/api/auth/login", request);

        if (!response.IsSuccessStatusCode)
            return false;

        var result = await response.Content.ReadFromJsonAsync<LoginResponse>();
        _token = result?.Token;

        if (!string.IsNullOrEmpty(_token))
        {
            _httpClient.DefaultRequestHeaders.Authorization =
                new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _token);
        }

        return !string.IsNullOrEmpty(_token);
    }

    /// <summary>
    /// Insert a vector
    /// </summary>
    public async Task<bool> InsertVectorAsync(string id, float[] vector, Dictionary<string, object>? metadata = null)
    {
        var request = new
        {
            id,
            vector = vector.ToList(),
            metadata
        };

        var response = await _httpClient.PostAsJsonAsync("/api/vectors", request);
        return response.IsSuccessStatusCode;
    }

    /// <summary>
    /// Batch insert vectors
    /// </summary>
    public async Task<int> BatchInsertAsync(List<(string id, float[] vector, Dictionary<string, object>? metadata)> vectors)
    {
        var request = new
        {
            vectors = vectors.Select(v => new
            {
                id = v.id,
                vector = v.vector.ToList(),
                metadata = v.metadata
            }).ToList()
        };

        var response = await _httpClient.PostAsJsonAsync("/api/vectors/batch", request);
        
        if (!response.IsSuccessStatusCode)
            return 0;

        var result = await response.Content.ReadFromJsonAsync<BatchInsertResponse>();
        return result?.Inserted ?? 0;
    }

    /// <summary>
    /// Search for similar vectors (KNN)
    /// </summary>
    public async Task<List<SearchResult>> SearchAsync(float[] queryVector, int k = 10, string metric = "cosine")
    {
        var request = new
        {
            vector = queryVector.ToList(),
            k,
            metric
        };

        var response = await _httpClient.PostAsJsonAsync("/api/search/knn", request);
        
        if (!response.IsSuccessStatusCode)
            return new List<SearchResult>();

        var result = await response.Content.ReadFromJsonAsync<SearchResponse>();
        return result?.Results ?? new List<SearchResult>();
    }

    /// <summary>
    /// Get vector by ID
    /// </summary>
    public async Task<VectorResponse?> GetVectorAsync(string id)
    {
        var response = await _httpClient.GetAsync($"/api/vectors/{id}");
        
        if (!response.IsSuccessStatusCode)
            return null;

        return await response.Content.ReadFromJsonAsync<VectorResponse>();
    }

    /// <summary>
    /// Delete a vector
    /// </summary>
    public async Task<bool> DeleteVectorAsync(string id)
    {
        var response = await _httpClient.DeleteAsync($"/api/vectors/{id}");
        return response.IsSuccessStatusCode;
    }

    /// <summary>
    /// Create an index
    /// </summary>
    public async Task<bool> CreateIndexAsync(string name, string table, string column, 
        string indexType = "hnsw", string metric = "cosine", Dictionary<string, object>? parameters = null)
    {
        var request = new
        {
            name,
            table,
            column,
            indexType,
            metric,
            parameters
        };

        var response = await _httpClient.PostAsJsonAsync("/api/indexes", request);
        return response.IsSuccessStatusCode;
    }

    /// <summary>
    /// List all indexes
    /// </summary>
    public async Task<List<IndexResponse>> ListIndexesAsync()
    {
        var response = await _httpClient.GetAsync("/api/indexes");
        
        if (!response.IsSuccessStatusCode)
            return new List<IndexResponse>();

        var result = await response.Content.ReadFromJsonAsync<IndexListResponse>();
        return result?.Indexes ?? new List<IndexResponse>();
    }
}

// Response models
public record LoginResponse(string Token, string RefreshToken, DateTime ExpiresAt);
public record BatchInsertResponse(int Inserted, string Message);
public record SearchResponse(List<SearchResult> Results, int TotalResults, double QueryTimeMs);
public record SearchResult(string Id, float Distance, float Similarity, Dictionary<string, object>? Metadata);
public record VectorResponse(string Id, List<float> Vector, Dictionary<string, object>? Metadata, DateTime? CreatedAt, DateTime? UpdatedAt);
public record IndexResponse(string Name, string Type, string Metric, long VectorCount, DateTime CreatedAt, string Status);
public record IndexListResponse(List<IndexResponse> Indexes);

/// <summary>
/// Example usage
/// </summary>
public class Program
{
    public static async Task Main(string[] args)
    {
        var client = new SfvectorClient("http://localhost:5000");

        // Login
        Console.WriteLine("Logging in...");
        var loginSuccess = await client.LoginAsync("admin", "password");
        if (!loginSuccess)
        {
            Console.WriteLine("Login failed!");
            return;
        }
        Console.WriteLine("✓ Logged in successfully");

        // Generate random vectors (simulating embeddings)
        var random = new Random();
        float[] GenerateVector(int dim) => 
            Enumerable.Range(0, dim).Select(_ => (float)random.NextDouble()).ToArray();

        // Insert vectors
        Console.WriteLine("\nInserting vectors...");
        await client.InsertVectorAsync("doc1", GenerateVector(1536), new Dictionary<string, object>
        {
            ["title"] = "Introduction to Machine Learning",
            ["category"] = "AI"
        });
        await client.InsertVectorAsync("doc2", GenerateVector(1536), new Dictionary<string, object>
        {
            ["title"] = "Deep Learning Fundamentals",
            ["category"] = "AI"
        });
        await client.InsertVectorAsync("doc3", GenerateVector(1536), new Dictionary<string, object>
        {
            ["title"] = "Database Optimization",
            ["category"] = "Database"
        });
        Console.WriteLine("✓ Inserted 3 vectors");

        // Batch insert
        Console.WriteLine("\nBatch inserting vectors...");
        var batchVectors = Enumerable.Range(4, 10)
            .Select(i => (
                id: $"doc{i}",
                vector: GenerateVector(1536),
                metadata: new Dictionary<string, object> { ["title"] = $"Document {i}", ["category"] = "General" }
            ))
            .ToList();

        var inserted = await client.BatchInsertAsync(batchVectors);
        Console.WriteLine($"✓ Batch inserted {inserted} vectors");

        // Search
        Console.WriteLine("\nSearching for similar vectors...");
        var queryVector = GenerateVector(1536);
        var results = await client.SearchAsync(queryVector, k: 5, metric: "cosine");

        Console.WriteLine($"Found {results.Count} similar vectors:");
        foreach (var result in results)
        {
            Console.WriteLine($"  - {result.Id}: similarity={result.Similarity:F3}, distance={result.Distance:F3}");
        }

        // Get specific vector
        Console.WriteLine("\nRetrieving vector 'doc1'...");
        var vector = await client.GetVectorAsync("doc1");
        if (vector != null)
        {
            Console.WriteLine($"✓ Retrieved vector: {vector.Id}");
            Console.WriteLine($"  Dimension: {vector.Vector.Count}");
            Console.WriteLine($"  Metadata: {JsonSerializer.Serialize(vector.Metadata)}");
        }

        // Create index
        Console.WriteLine("\nCreating HNSW index...");
        var indexCreated = await client.CreateIndexAsync(
            name: "vectors_hnsw",
            table: "VectorsTest_FAISS",
            column: "vector",
            indexType: "hnsw",
            metric: "cosine",
            parameters: new Dictionary<string, object>
            {
                ["M"] = 32,
                ["efConstruction"] = 200
            }
        );
        Console.WriteLine(indexCreated ? "✓ Index created" : "✗ Index creation failed");

        // List indexes
        Console.WriteLine("\nListing indexes...");
        var indexes = await client.ListIndexesAsync();
        Console.WriteLine($"Found {indexes.Count} indexes:");
        foreach (var index in indexes)
        {
            Console.WriteLine($"  - {index.Name}: {index.Type} ({index.VectorCount} vectors)");
        }

        Console.WriteLine("\n✓ All operations completed successfully!");
    }
}

File: api/examples/client-example.js
// JavaScript/TypeScript client example for sfvector API

class SfvectorClient {
    constructor(baseUrl) {
        this.baseUrl = baseUrl;
        this.token = null;
    }

    /**
     * Login and get JWT token
     */
    async login(username, password) {
        const response = await fetch(`${this.baseUrl}/api/auth/login`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ username, password })
        });

        if (!response.ok) {
            throw new Error('Login failed');
        }

        const data = await response.json();
        this.token = data.token;
        return data;
    }

    /**
     * Get authorization headers
     */
    getHeaders() {
        return {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${this.token}`
        };
    }

    /**
     * Insert a vector
     */
    async insertVector(id, vector, metadata = null) {
        const response = await fetch(`${this.baseUrl}/api/vectors`, {
            method: 'POST',
            headers: this.getHeaders(),
            body: JSON.stringify({ id, vector, metadata })
        });

        if (!response.ok) {
            throw new Error(`Insert failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Batch insert vectors
     */
    async batchInsert(vectors) {
        const response = await fetch(`${this.baseUrl}/api/vectors/batch`, {
            method: 'POST',
            headers: this.getHeaders(),
            body: JSON.stringify({ vectors })
        });

        if (!response.ok) {
            throw new Error(`Batch insert failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Search for similar vectors (KNN)
     */
    async search(queryVector, k = 10, metric = 'cosine') {
        const response = await fetch(`${this.baseUrl}/api/search/knn`, {
            method: 'POST',
            headers: this.getHeaders(),
            body: JSON.stringify({ vector: queryVector, k, metric })
        });

        if (!response.ok) {
            throw new Error(`Search failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Get vector by ID
     */
    async getVector(id) {
        const response = await fetch(`${this.baseUrl}/api/vectors/${id}`, {
            method: 'GET',
            headers: this.getHeaders()
        });

        if (!response.ok) {
            if (response.status === 404) return null;
            throw new Error(`Get vector failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Delete a vector
     */
    async deleteVector(id) {
        const response = await fetch(`${this.baseUrl}/api/vectors/${id}`, {
            method: 'DELETE',
            headers: this.getHeaders()
        });

        return response.ok;
    }

    /**
     * List vectors with pagination
     */
    async listVectors(page = 1, pageSize = 50) {
        const response = await fetch(
            `${this.baseUrl}/api/vectors?page=${page}&pageSize=${pageSize}`,
            {
                method: 'GET',
                headers: this.getHeaders()
            }
        );

        if (!response.ok) {
            throw new Error(`List vectors failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Create an index
     */
    async createIndex(name, table, column, indexType = 'hnsw', metric = 'cosine', parameters = null) {
        const response = await fetch(`${this.baseUrl}/api/indexes`, {
            method: 'POST',
            headers: this.getHeaders(),
            body: JSON.stringify({ name, table, column, indexType, metric, parameters })
        });

        if (!response.ok) {
            throw new Error(`Create index failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * List all indexes
     */
    async listIndexes() {
        const response = await fetch(`${this.baseUrl}/api/indexes`, {
            method: 'GET',
            headers: this.getHeaders()
        });

        if (!response.ok) {
            throw new Error(`List indexes failed: ${response.statusText}`);
        }

        return await response.json();
    }

    /**
     * Get index statistics
     */
    async getIndexStats(name) {
        const response = await fetch(`${this.baseUrl}/api/indexes/${name}/stats`, {
            method: 'GET',
            headers: this.getHeaders()
        });

        if (!response.ok) {
            if (response.status === 404) return null;
            throw new Error(`Get index stats failed: ${response.statusText}`);
        }

        return await response.json();
    }
}

// Example usage
async function main() {
    const client = new SfvectorClient('http://localhost:5000');

    try {
        // Login
        console.log('Logging in...');
        await client.login('admin', 'password');
        console.log('✓ Logged in successfully');

        // Helper function to generate random vectors
        const generateVector = (dim) => 
            Array(dim).fill(0).map(() => Math.random());

        // Insert vectors
        console.log('\nInserting vectors...');
        await client.insertVector('doc1', generateVector(1536), {
            title: 'Introduction to Machine Learning',
            category: 'AI'
        });
        await client.insertVector('doc2', generateVector(1536), {
            title: 'Deep Learning Fundamentals',
            category: 'AI'
        });
        console.log('✓ Inserted 2 vectors');

        // Batch insert
        console.log('\nBatch inserting vectors...');
        const batchVectors = Array.from({ length: 10 }, (_, i) => ({
            id: `doc${i + 3}`,
            vector: generateVector(1536),
            metadata: {
                title: `Document ${i + 3}`,
                category: 'General'
            }
        }));

        const batchResult = await client.batchInsert(batchVectors);
        console.log(`✓ Batch inserted ${batchResult.inserted} vectors`);

        // Search
        console.log('\nSearching for similar vectors...');
        const queryVector = generateVector(1536);
        const searchResult = await client.search(queryVector, 5, 'cosine');

        console.log(`Found ${searchResult.totalResults} similar vectors (${searchResult.queryTimeMs.toFixed(2)}ms):`);
        searchResult.results.forEach(result => {
            console.log(`  - ${result.id}: similarity=${result.similarity.toFixed(3)}, distance=${result.distance.toFixed(3)}`);
        });

        // Get specific vector
        console.log('\nRetrieving vector "doc1"...');
        const vector = await client.getVector('doc1');
        if (vector) {
            console.log(`✓ Retrieved vector: ${vector.id}`);
            console.log(`  Dimension: ${vector.vector.length}`);
            console.log(`  Metadata:`, vector.metadata);
        }

        // List vectors
        console.log('\nListing first page of vectors...');
        const vectorList = await client.listVectors(1, 10);
        console.log(`Total vectors: ${vectorList.total}, Page: ${vectorList.page}/${vectorList.totalPages}`);
        vectorList.vectors.forEach(v => {
            console.log(`  - ${v.id}`);
        });

        // Create index
        console.log('\nCreating HNSW index...');
        const indexResult = await client.createIndex(
            'vectors_hnsw',
            'VectorsTest_FAISS',
            'vector',
            'hnsw',
            'cosine',
            { M: 32, efConstruction: 200 }
        );
        console.log('✓ Index created:', indexResult.name);

        // List indexes
        console.log('\nListing indexes...');
        const indexes = await client.listIndexes();
        console.log(`Found ${indexes.indexes.length} indexes:`);
        indexes.indexes.forEach(index => {
            console.log(`  - ${index.name}: ${index.type} (${index.vectorCount} vectors)`);
        });

        console.log('\n✓ All operations completed successfully!');

    } catch (error) {
        console.error('Error:', error.message);
    }
}

// Run in Node.js
if (typeof module !== 'undefined' && module.exports) {
    // For Node.js, you'll need to install 'node-fetch':
    // npm install node-fetch
    const fetch = require('node-fetch');
    main();
}

// Export for browser use
if (typeof window !== 'undefined') {
    window.SfvectorClient = SfvectorClient;
}
